vowpalwabbit/interactions.cc

```diff
@@ -18,140 +21,118 @@ using namespace VW::config;
+{
+  for (auto value : feature_spaces[term].values) { func(value); }
+}
 
-// returns number of new features that will be generated for example and sum of their squared values
-void eval_count_of_generated_ft(bool permutations, const std::vector<std::vector<namespace_index>>& interactions,
-    const std::array<features, NUM_NAMESPACES>& feature_spaces, size_t& new_features_cnt, float& new_features_value)
+template <typename FuncT>
+void for_each_value(const std::array<features, NUM_NAMESPACES>& feature_spaces, extent_term term, const FuncT& func)
 {
-  new_features_cnt = 0;
-  new_features_value = 0.;
+  feature_spaces[term.first].foreach_feature_for_hash(
+      term.second, [&](features::const_audit_iterator it) { func(it.value()); });
+}
 
-  v_array<float> results;
+float calc_sum_ft_squared_for_term(const std::array<features, NUM_NAMESPACES>& feature_spaces, extent_term term)
+{
+  float sum_feat_sq_in_inter = 0.f;
+  feature_spaces[term.first].foreach_feature_for_hash(
+      term.second, [&](features::const_audit_iterator it) { sum_feat_sq_in_inter += it.value() * it.value(); });
+  return sum_feat_sq_in_inter;
+}
 
-  if (permutations)
+float calc_sum_ft_squared_for_term(const std::array<features, NUM_NAMESPACES>& feature_spaces, namespace_index term)
+{
+  return feature_spaces[term].sum_feat_sq;
+}
+
+template <typename InteractionTermT>
+float calculate_count_and_sum_ft_sq_for_permutations(const std::array<features, NUM_NAMESPACES>& feature_spaces,
+    const std::vector<std::vector<InteractionTermT>>& interactions)
+{
+  float sum_feat_sq_in_inter_outer = 0.;
+  for (const auto& interaction : interactions)
   {
-    // just multiply precomputed values for all namespaces
-    for (const auto& inter : interactions)
-    {
-      size_t num_features_in_inter = 1;
-      float sum_feat_sq_in_inter = 1.;
+    float sum_feat_sq_in_inter = 1.;
+    for (auto term : interaction) { sum_feat_sq_in_inter *= calc_sum_ft_squared_for_term(feature_spaces, term); }
+    sum_feat_sq_in_inter_outer += sum_feat_sq_in_inter;
+  }
 
-      for (namespace_index ns : inter)
-      {
-        num_features_in_inter *= feature_spaces[ns].size();
-        sum_feat_sq_in_inter *= feature_spaces[ns].sum_feat_sq;
-        // If there are no features, then we don't want to accumulate the default value of 1.0, so we zero out here.
-        if (num_features_in_inter == 0) { sum_feat_sq_in_inter = 0; }
-      }
+  return sum_feat_sq_in_inter_outer;
+}
 
-      new_features_cnt += num_features_in_inter;
-      new_features_value += sum_feat_sq_in_inter;
-    }
-  }
-  else  // case of simple combinations
+template <typename InteractionTermT>
+float calculate_count_and_sum_ft_sq_for_combinations(const std::array<features, NUM_NAMESPACES>& feature_spaces,
+    const std::vector<std::vector<InteractionTermT>>& interactions)
+{
+  std::vector<float> results;
```

Comment:
memory_tree.cc uses get_total_sum_feat_sq() in its predict() and learn() functions which also ends up here.

Reply:
The size of this vector is the order of the interaction. So for quadratics it is 2. I doubt reusing a buffer of size 2 or 3 (up to maybe 4 in some cases?) is going to be out perform. I think the composability and maintainability cost of requiring an existing buffer to be passed to this function are too high for the questionable benefit.

Reply:
The dependency between size of allocation and cost of allocation is not easily quantifiable.  It depends on implementation of allocator, system load (cpu and memory).

Reply:
This looks like premature optimization, if the benchmarks show a regression we should profile the PR and see where the pain point is instead of sacrificing code quality without any data

Reply:
Unlike managed languages, heap allocations in C++ is expensive and has a large variance depending on operating conditions.  Getting an accurate benchmark on memory allocations is tricky.  It depends on a) allocation patterns in the process b) allocator that's used by the application c)  system load (both cpu and memory).  Many of these only surface in the field and debugging perf issues related to memory allocations is difficult.

In addition to perf issues, VW does not deal well with allocation failures in general so avoiding allocations also leads to more stable runtime behavior.

VW has always avoided allocations in steady state for these reasons.
 
This is common practice in applications that are performance sensitive (For example game dev or fintech).  

There are many good things here so I created a tracking item to unblock this PR.